protocols/swap/reports/SECURITY.md

```diff
@@ -1,175 +1,173 @@
-  No functions are payable and thus stuck ether due to improper function calls is not possible.
-  By inspection, in all branches of this function, swap, and safeTransferAny/transferAny, either the trade succeeds or an exception is generated by a ‘require’. At no point can these functions return false or fail silently without throwing.
-- All transfers in the execution function are between the maker and taker or a maker and affiliate; at no point is a token transfer performed using the Swap contract’s address. Tokens can however be sent to the contract by accident, and will remain stuck as there is no withdrawal mechanism for any of the supported token standards. Per V1 Exchange contract (0x8fd3121013a07c57f0d69646e86e7a4880b467b7), there is 0 ether and < \$.10 stored. While we recommend considering a withdrawal backdoor, it introduces complexity that may be undesirable from a business perspective.
-- **This invariant currently holds as-is.**
+- **This invariant holds as-is.**
 
-#### 2. Affiliate trades supply appropriate fees
+### The `swap` function will only either successfully settle an order or revert.
 
-- safeTransferFrom is used for ERC-721 and transferFrom is called with affiliate.param as amount, transfer works by manual verification of transfer flow and testing; see also the original Swap audit report. A require is not required for safeTransferFrom as the contract signature guarantees there is no return parameter.
-- **This invariant currently holds as-is.**
+- By manual inspection and testing, after checks for expiry (line 87), nonce validity (lines 91, 95), sender validity (lines 102:121), and signer and signature validity (lines 124:144), the swap completes, barring issues on the underlying token contracts. Each of these checks will revert with a reason in a failure case.
+- A `require` is used in the ERC-20 case to handle tokens where a bad transfer may fail silently without reversions. This is the only external call other than safeTransferFrom, for the ERC-721 case, which has no return value and therefore cannot fail silently.
+- **This invariant holds as-is.**
 
-#### 3. All flows are identical (no calls to underlying token contract) until the internal transferToken call.
+### An order cannot be settled at any time beyond the `expiry` specified on the order.
 
-- In the ERC-20 case, reverts will occur on the underlying transferFrom implemented methods. Applications using this contract may want to do pre-checks on known tokens to do better user experience.
-- **This invariant currently holds as-is.**
+- The `expiry` of each order is checked in the first line of the `swap` function, comparing the `expiry` to `block.timestamp` and throwing if at or beyond it.
+- **This invariant holds as-is.**
 
-#### 4. Signatures cannot be forged or duplicated.
+### Signatures cannot be forged or duplicated.
```

Comment:
Feels like the following 2 invariants could potentially be merged? They certainly have overlap.
1) Signatures cannot be forged or duplicated.
2) Orders can only be successfully taken once.
Maybe 2 is a subset of 1?

Reply:
Updated the first one to be more signature-centric.